### PR TITLE
BLIP: enable generation tests

### PR DESCRIPTION
**Pull request by zucchini-nlp**
# What does this PR do?

Enables generation tests for BLIP models, except BLIP-1 (turned out to be a bit harder). I changed the generation tests to use `modelTest.input_name` as BLIP is the only model that uses pixel values as main input and thus checking generated text length's will always fail. 

I tried to get rid of custom generate for these models, but that opened a Pandora box so I think better not waste time on an old model and maintain it for a while, until the model gets deprecated. But still I did some changes so we don't need to add extra `bos` at the beginning and now the decoder-based BLIP models return full text at output. Encoder-decoder based models return only generated text, which is consistent with what an LLM should do

### CODE REVIEW

**Comment by zucchini-nlp** [comment]
@gante I tried to force `input_ids` always and I found another corner case with ~~Whisper which expects `input_features` to be the main input for shape checking~~. I would very much love to make BLIP standard and maybe I'll make so in v5, because it will break a whole lot of things.

OMG, I found an option while writing this reply, whisper and the other audio model are encoder-decoder so we can make it work by getting main input in decoder-only models. Just before the check happens, in the same indent block  :)

**Review by ArthurZucker — COMMENTED** [review]
AH actually we might need / want to force `return_dict` to TRUE, to avoid all the if elses

**Comment by gante** [comment]
> OMG, I found an option while writing this reply, whisper and the other audio model are encoder-decoder so we can make it work by getting main input in decoder-only models. Just before the check happens, in the same indent block :)
If you agree, I'll make the change hehe

if it works, sounds good! (make sure to leave a comment)

**Comment by zucchini-nlp** [comment]
@gante requesting re-review, since the `input-name` was merged as a separate PR. I rebased `main` and ran tests again

File: src/transformers/models/blip_2/modeling_blip_2.py [hunk 1]
```diff
@@ -1768,11 +1768,12 @@ def forward(
...
                 output_hidden_states=output_hidden_states,
-                return_dict=return_dict,
+                return_dict=True,  # toggle for easier access to loss/logits below
                 labels=labels,
             )
-            loss = outputs.loss if return_dict else outputs[0]
-            logits = outputs.logits if return_dict else outputs[1]
+            loss = outputs.loss
+            logits = outputs.logits
+            outputs = outputs.to_tuple() if not return_dict else outputs
```
Comment (ArthurZucker) [inline src/transformers/models/blip_2/modeling_blip_2.py:1776]: outputs will have loss and logits twice there

Comment (ArthurZucker) [reply src/transformers/models/blip_2/modeling_blip_2.py:1776]: (this was probably already a bug)

File: tests/models/instructblip/test_modeling_instructblip.py [hunk 2]
```diff
@@ -531,6 +550,214 @@ def test_model_from_pretrained(self):
...
+            "reformer",
+            "jamba",
+            "mamba",
+            "xlnet",
+            "zamba",
+        )
+        has_standard_cache = not any(
+            model_name in config.__class__.__name__.lower() for model_name in models_without_standard_cache
+        )
+        if has_standard_cache:
```
Comment (ArthurZucker) [inline tests/models/instructblip/test_modeling_instructblip.py:654]: when we overwrite we don't need that no?

**Review by ArthurZucker — APPROVED** [review]
Thanks for cleaning up

File: src/transformers/models/blip_2/modeling_blip_2.py (hunk 1 above)
Comment (zucchini-nlp) [reply src/transformers/models/blip_2/modeling_blip_2.py:1776]: yep, in general i don't like that we return it as this and would better return unwrapped lm outputs. But we can't prob just delete it for BC reasons
